swift_package/BrainFlow/Sources/BrainFlowBindings/DataFilter.swift

```diff
@@ -0,0 +1,650 @@
+        
+        if (newSize <= 0) {
+            throw BrainFlowException ("Invalid data size", .INVALID_ARGUMENTS_ERROR)
+        }
+        
+        var downsampledData = [Double](repeating: 0.0, count: Int(newSize))
+        var cData = data
+        let errorCode = perform_downsampling (&cData, dataLen, period, operation.rawValue, &downsampledData)
+        try checkErrorCode("Failed to perform downsampling", errorCode)
+
+        return downsampledData
+    }
+    
+    /**
+     * removes noise using notch filter in-place
+     */
+    static func removeEnvironmentalNoise (data: inout [Double], samplingRate: Int32, noiseType: NoiseTypes) throws {
+        let dataLen = Int32(data.count)
+        let errorCode = remove_environmental_noise (&data, dataLen, samplingRate, noiseType.rawValue)
+        try checkErrorCode("Failed to remove noise", errorCode)
+    }
+
+    /**perform wavelet denoising
+
+    :param data: data to denoise
+    :type data: NDArray[Float64]
+    :param wavelet: use WaveletTypes enum
+    :type wavelet: int
+    :param decomposition_level: decomposition level
+    :type decomposition_level: int
+    :param wavelet_denoising: use WaveletDenoisingTypes enum
+    :type wavelet_denoising: int
+    :param threshold: use ThresholdTypes enum
+    :type threshold: int
+    :param extension_type: use WaveletExtensionTypes enum
+    :type extension_type: int
+    :param noise_level: use NoiseEstimationLevelTypes enum
+    :type noise_level: int
+    **/
+    static func performWaveletDenoising (data: inout [Double], wavelet: WaveletTypes, decompositionLevel: Int32,
+                                         waveletDenoising: WaveletDenoisingTypes = .SURESHRINK,
+                                         threshold: ThresholdTypes = .HARD,
+                                         extensionType: WaveletExtensionTypes = .SYMMETRIC, 
+                                         noiseLevel: NoiseEstimationLevelTypes = .FIRST_LEVEL) throws {
+        let dataLen = Int32(data.count)
+        let wvVal = wavelet.rawValue
+        let wvDenVal = waveletDenoising.rawValue
+        let thrVal = threshold.rawValue
+        let extVal = extensionType.rawValue
+        let noiseVal = noiseLevel.rawValue
+        
+        let errorCode = perform_wavelet_denoising (&data, dataLen, wvVal,
+                                                   decompositionLevel, wvDenVal, thrVal, extVal, noiseVal)
+        try checkErrorCode("Failed to perform denoising", errorCode)
+                                             
+        return
+    }
+    
+    /**
+     * perform wavelet transform
+     *
+     * @param wavelet supported vals:
+     *                db1..db15,haar,sym2..sym10,coif1..coif5,bior1.1,bior1.3,bior1.5,bior2.2,bior2.4,bior2.6,bior2.8,bior3.1,bior3.3,bior3.5
+     *                ,bior3.7,bior3.9,bior4.4,bior5.5,bior6.8
+     */
+    static func performWaveletTransform(data: [Double], wavelet: WaveletTypes,
```

Comment:
did you run wavelet test? did you compare original data and restored? the same for fft